homeassistant/components/http.py

```diff
@@ -106,413 +68,347 @@ def setup(hass, config):
+        """Register a folder to serve as a static path."""
+        from static import Cling
 
-    def set_cache_header(self):
-        """Add cache headers if not in development."""
-        if self.server.development:
-            return
+        headers = []
 
-        # 1 year in seconds
-        cache_time = 365 * 86400
+        if not self.development:
+            # 1 year in seconds
+            cache_time = 365 * 86400
 
-        self.send_header(
-            HTTP_HEADER_CACHE_CONTROL,
-            "public, max-age={}".format(cache_time))
-        self.send_header(
-            HTTP_HEADER_EXPIRES,
-            self.date_time_string(time.time()+cache_time))
+            headers.append({
+                'prefix': '',
+                HTTP_HEADER_CACHE_CONTROL:
+                "public, max-age={}".format(cache_time)
+            })
 
-    def set_session_cookie_header(self):
-        """Add the header for the session cookie and return session ID."""
-        if not self.authenticated:
-            return None
+        self.register_wsgi_app(url_root, Cling(path, headers=headers))
 
-        session_id = self.get_cookie_session_id()
+    def register_wsgi_app(self, url_root, app):
+        """Register a path to serve a WSGI app."""
+        if url_root in self.extra_apps:
+            _LOGGER.warning("Url root '%s' is being overwritten", url_root)
 
-        if session_id is not None:
-            self.server.sessions.extend_validation(session_id)
-            return session_id
+        self.extra_apps[url_root] = app
 
-        self.send_header(
-            'Set-Cookie',
-            '{}={}'.format(SESSION_KEY, self.server.sessions.create())
+    def start(self):
+        """Start the wsgi server."""
+        from eventlet import wsgi
+        import eventlet
+
+        sock = eventlet.listen((self.server_host, self.server_port))
+        if self.ssl_certificate:
+            eventlet.wrap_ssl(sock, certfile=self.ssl_certificate,
+                              keyfile=self.ssl_key, server_side=True)
+        wsgi.server(sock, self)
+
+    def dispatch_request(self, request):
+        """Handle incoming request."""
+        from werkzeug.exceptions import (
+            MethodNotAllowed, NotFound, BadRequest, Unauthorized,
         )
+        from werkzeug.routing import RequestRedirect
 
-        return session_id
-
-    def verify_session(self):
-        """Verify that we are in a valid session."""
-        return self.get_cookie_session_id() is not None
+        with request:
+            adapter = self.url_map.bind_to_environ(request.environ)
+            try:
+                endpoint, values = adapter.match()
+                resp = self.views[endpoint].handle_request(request, **values)
+
+                if resp.is_streamed or 200 > resp.status_code >= 300 or \
+                   'gzip' not in request.accept_encodings:
+                    return resp
+
+                resp.headers.extend([('Content-Encoding', 'gzip'),
```

Comment:
I'm glad having you people around. Thanks.

For the bootstrap api call it can  change from 30kB to 5kB (after that it's
streaming so all good).

On the wiki you linked there is this

> Another suggested approach is to disable HTTP compression whenever the
referrer header indicates a cross-site request, or when the header is not
present.[4][5] This approach allows effective mitigation of the attack
without losing functionality, only incurring a performance penalty on
affected requests.

Would that be an option? Or maybe we should do the bootstrap over
websockets too so it will not be an issue the wsgi handler has to solve.

On Fri, May 20, 2016, 08:00 Martin Hjelmare <notifications@github.com>
wrote:

> In homeassistant/components/http.py
> <https://github.com/home-assistant/home-assistant/pull/2063#discussion_r64055286>
> :
>
> > -        return session_id
> > -
> > -    def verify_session(self):
> > -        """Verify that we are in a valid session."""
> > -        return self.get_cookie_session_id() is not None
> > +        with request:
> > +            adapter = self.url_map.bind_to_environ(request.environ)
> > +            try:
> > +                endpoint, values = adapter.match()
> > +                resp = self.views[endpoint].handle_request(request, **values)
> > +
> > +                if resp.is_streamed or 200 > resp.status_code >= 300 or \
> > +                   'gzip' not in request.accept_encodings:
> > +                    return resp
> > +
> > +                resp.headers.extend([('Content-Encoding', 'gzip'),
>
> There's a quote that I like to tout: "Safety and precision first, speed
> will follow."
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/2063/files/95baff60b30d051c1ca0a2a572ddb9b062d2e8de#r64055286>
>